design/ceph/multus-network.md

```diff
@@ -0,0 +1,110 @@
+# Multus networking integration with Ceph
+
+We have already explored and explained the benefit of multi-homed networking, so this document will not rehearse that but simply focus on the implementation for the Ceph backend.
+If you are interested in learning more about multi-homed networking you can read the [design documentation on that matter](../core/multi-homed-cluster.md).
+
+To make the story short, [Multus](https://github.com/intel/multus-cni) should allow us to get the same performance benefit as `HostNetworking` by increasing the security.
+Using `HostNetworking` results in exposing **all** the network interfaces (the entire stack) of the host inside the container where Multus allows you to pick the one you want.
+Also, this removes the need of privileged containers (required for `HostNetworking`).
+
+## Proposed CRD changed
+
+We already have a `network` CRD property, which looks like:
+
+```yaml
+network:
+  provider:
+  selectors:
+```
+
+We will expand the `selectors` with the following two hardcoded keys:
+
+```yaml
+selectors:
+  public:
+  cluster:
+```
+
+Each selector represents a [`NetworkAttachmentDefinition`](https://github.com/intel/multus-cni/blob/master/doc/quickstart.md#storing-a-configuration-as-a-custom-resource) object in Multus.
+At least, one must be provided and by default they will represent:
+
+- `public`: data daemon public network (binds to `public_network` Ceph option).
+- `cluster`: data daemon cluster network (binds to `cluster_network` Ceph option).
+
+If only `public` is set then `cluster` will take the value of `public`.
+
+## Multus supported configuration
+
+### Interface type
+
+As part of the CNI spec, Multus supports several [interface types](https://github.com/containernetworking/plugins#main-interface-creating).
+Rook will naturally support any of them as they don't fundamentally change the working behavior.
+
+### IPAM type
+
+This is where things get more complex.
+Currently there are [three different IPAM](https://github.com/containernetworking/plugins#ipam-ip-address-allocation) solutions available.
+
+As part of my research I have found that the following IPAM types are not good candidates:
+
+- host-local: Maintains a local database of allocated IPs, this only works on a per host basis, so not suitable for a distributed environment since we will end up with IP collision
+To fix this, the [whereabouts project](https://github.com/dougbtv/whereabouts) looks promising but is not officially supported.
+- static: Allocate a static IPv4/IPv6 addresses to container and it's useful in debugging purpose.
+This cannot at scale because this means we will have to allocate IPs for **all** the daemon so it's not scalable.
+
+## Monitors and OSDs implementation challenges
+
+The Ceph monitors only need access to the public network.
+The OSDs needs access to both public and cluster networks.
+
+Monitors requirements so far are the following:
+
+- Predictable IP addresses
```

Comment:
If multus helps us with the ip address bindings in the pod for all the benefits of security and performance, it seems like the K8s service could still give us the stable endpoint. Is there a reason Multus won't work with the K8s services?

Reply:
For now, multus does not interwork with Kubernetes service object/component (i.e. kube-proxy). In addition, kube-proxy IP service is always in container host, hence some multus interfaces (like SR-IOV interface) send/recv traffic without container host routing/kube-proxy iptables as well.

Multus team is now designing service abstraction for multus. So it is pretty TBD item.

Reply:
@leseb We are blocked from moving to Multus on this issue, or is there another approach we can use for the mons?

Reply:
Yes, we are, not as far as I know. I need to jump into the network plumbing group weekly meeting to ask questions... Probably somewhere after KubeCon.